wazo_confd/plugins/line/middleware.py

```diff
@@ -119,3 +122,99 @@ def delete(self, line_id, tenant_uuids, recursive=False):
+        for name, value in form.items():
+            setattr(model, name, value)
+        self._service.edit(model, updated_fields=updated_fields, **kwargs)
+
+    def find_updated_fields(self, model, form):
+        updated_fields = []
+        for name, value in form.items():
+            try:
+                if getattr(model, name) != value:
+                    updated_fields.append(name)
+            except AttributeError:
+                pass
+        return updated_fields
+
+    def update(self, line_id, body, tenant_uuid, tenant_uuids):
+        model = self._service.get(line_id, tenant_uuids=tenant_uuids)
+
+        endpoint_custom_body = body.pop('endpoint_custom', None)
+        endpoint_sccp_body = body.pop('endpoint_sccp', None)
+        endpoint_sip_body = body.pop('endpoint_sip', None)
+        extension_bodies = body.pop('extensions', None) or []
+
+        if endpoint_sip_body is not None:
+            if not model.endpoint_sip:
+                raise InputError(
+                    "There is already an endpoint associated to the line that is not of type SIP. Cannot update the line with another endpoint type"
+                )
+            endpoint_sip_middleware = self._middleware_handle.get('endpoint_sip')
+            endpoint_sip_middleware.update(
+                endpoint_sip_body['uuid'], endpoint_sip_body, tenant_uuids
+            )
+
+        elif endpoint_sccp_body is not None:
+            if not model.endpoint_sccp:
+                raise InputError(
+                    "There is already an endpoint associated to the line that is not of type SCCP. Cannot update the line with another endpoint type"
+                )
+            endpoint_sccp_middleware = self._middleware_handle.get('endpoint_sccp')
+            endpoint_sccp_middleware.update(
+                endpoint_sccp_body['uuid'], endpoint_sccp_body, tenant_uuids
+            )
+
+        elif endpoint_custom_body is not None:
+            if not model.endpoint_custom:
+                raise InputError(
+                    "There is already an endpoint associated to the line that is not of type custom. Cannot update the line with another endpoint type"
+                )
+            endpoint_custom_middleware = self._middleware_handle.get('endpoint_custom')
+            endpoint_custom_middleware.update(
+                endpoint_custom_body['uuid'], endpoint_custom_body, tenant_uuids
+            )
+
+        if extension_bodies:
+            line_extension_middleware = self._middleware_handle.get('line_extension')
+            existing_extensions = {ext.id: ext for ext in model.extensions}
+
+            extensions_body_to_be_updated = {}
+            extensions_body_to_be_created = []
+            extensions_body_to_be_deleted = []
+            for extension_body in extension_bodies:
+                try:
+                    extensions_body_to_be_updated[extension_body['id']] = extension_body
```

Comment:
Is the following case handled somewhere?
```javascript
{'exten': '1004', 'context': 'default', 'id': null}
```

Reply:
I updated the schema and added a test